### ISSUE TITLE
Extremely slow training speed for resizing 

### ISSUE THREAD

**Issue by FishWoWater** [hbertiche/PBNS#4]
Hi!  Thanks for your excellent project!

When I was training PBNS_resize on provided samples, the training speed seemed to be very slow (about 5mins / epoch on 3090 GPU, which means it will take 4 days to finish 1000 epochs)

The GPU util is very low also(3~5%).

Do you have any idea on how to improve the speed?  I guess the bottleneck can be the data loading or loss calculation?

Any help would be appreciated.

**Comment by hbertiche** [comment]
Hi,

thanks for showing interest in our work!

as you mention, the bottleneck here is not the network itself, but the computation of the loss, more specifically, the collision loss, which requires nearest neighbour matching, and it is performed by the CPU,

we already considered this, you'll see there is an alternative file for the losses, Losses.py and LossesRay.py, the former uses the ray library for multi-processing, this will parallelize the collision detection,
the increase in performance will be more noticeable for big batches and machines with high CPU count,

to use this alternative implementation, the changes required are minimal,
in train.py script, comment line #12 and uncomment line #14

I hope I could answer your question

**Comment by FishWoWater** [comment]
@hbertiche  Thanks for your quick reply! 

Yes, I have seen the alternative parallel loss version and it does speed up the training a lot. 

By the way, I'd like to extend your PBNS_Resizer to an arbitrary pose, which means we are learning shape-blend-shapes for some specific pose(only one, not only). I am considering about its feasibility and could you give me some insights?

Thanks, nice and helpful project really

**Comment by hbertiche** [comment]
what you propose sounds like PBNS_Resizer but with a different rest pose (instead of the usual T-pose for SMPL),
truth is our methodology already has a rest pose different from the T-pose, so it is perfectly possible to do so,

nonetheless, there are some issues to take into account regarding blend shapes,
blend shapes are defined in T-pose, meaning that applying such deformations in different poses will lead to inconsistent body geometries,
to solve this, we "posed" the blend shapes in our model, by rotating each vertex blend shapes according to our new rest pose, 
posing the body mesh means computing a 4x4 transformation matrix for each body vertex from the joint transformations and the blend weights, these per-vertex 4x4 transformations consist in a rotation and translation, then, we used the rotation part of these matrices to rotate each vertex blend shapes (blendshapes are like 6890x3x10, which can be understood as 6890 matrices like 3x10),
so it is a bit tricky, specially for challenging poses, but it is possible,

another possibility is to first shape the body and then pose it to the desire target pose (standard SMPL pipeline), nonetheless, for PBNS_Resizer, we first transfer blend shapes from the body to the outfit by proximity, and these blend shapes have to be consistent with what we re-defined as the new rest pose (since the outfit template will be aligned with this rest pose),
it is also important if you want to learn shape-blend-shapes from your specific pose (not in T-pose and then pose it),

I hope this was helpful!

**Comment by FishWoWater** [comment]
Thanks for your prompt and insightful reply!  You are such an expert in this field :+1: 

I am not sure I have fully caught your ideas and here is my understanding about this problem:

When we take T-pose as the rest pose, things are much simpler. What we need to learn is the displacement of the cloth, which can generalize to different body shapes. In your work, you decouple this learning into two parts:  a) obtain a corase cloth blend shapes by querying the body blend shapes by promixity, b) learn a displacement map for refinement(which can be expressed as the combination of serveral bases). 

But when it comes to some other pose as the rest pose, things become a bit challenging. I think there are two aspects of error: 1)  we obtain the cloth skinning weight by finding the closest body vertex and when we use this approximated cloth skin weight to transform the cloth, deviations can accumulate( especially for some loose clothing) 2) the standard shape-blend-shapes for the body(i.e. `shapedirs` in smpl) is defined in T-pose space,  and it can be different when we considering some other poses(actually SMPL introduces `posedirs` to correct this).  

You have mentioned two different strategies to address this limitations. 
1) We still align the cloth and the body in the T-pose, and then follow the standard SMPL pipeline to pose the cloth and body to some desired pose. In this solution,  both types of error exist and we expect the neural network to learn a powerful displacement to cover it.  This solution is straightforward but can be sub-optimal. 
2) We align the cloth and the body in the desired rest-pose(e.g. by arcsim or manually).  In this way,  the two types of error can be largely alleviated. But we have a new problem: how to define the shape-blend-shapes in the new pose. You mentioned the solution:  propagate the shape-blend-shapes from the T-pose to some desired pose by treating them as something like the vertices and use LBS. I think in this way we can still use the derived shape-blend-shapes as an initialization for the cloth (also setup by proximity). 

I am not sure I have understand your ideas correctly, but it does inspire me a lot. I have already implemented the simpler solution, the other seems more promising and I will try that if my understanding is correct. 

Thanks, Bertiche, wish you a happy day~

**Comment by FishWoWater** [comment]
By the way, it seems that current PBNS_Resizer can not well address different materials, since it's tricky to tune the weights of bending loss, edge loss and collision loss according to the properties of different materials.

**Comment by hbertiche** [comment]
I'm happy I could help you with my explanations =)

regarding the different materials for PBNS_Resizer, you are correct, it might be tricky to set a correct balance, but if done carefully, it is possible to obtain different cloth behaviours, just make sure not to go out of the range of values that ensure that the model works and converges,
this has to be done through trail and error

**Comment by qiqigit** [comment]
> @hbertiche Thanks for your quick reply!
> 
> Yes, I have seen the alternative parallel loss version and it does speed up the training a lot.

@FishWoWater 
Hi, I am wondering if you could share some insights on how to use ray library for parallelization.
-Did you make adjustments other than commenting line #12 and uncommenting line #14?
-How many CPUs did you use? 
-How big was your batch size?
-How long does it take to train after switching to LossesRay.py?
